cpp/src/arrow/types/struct.h

```diff
@@ -23,7 +23,137 @@
 #include <vector>
 
 #include "arrow/type.h"
+#include "arrow/types/primitive.h"
+#include "arrow/types/list.h"
 
-namespace arrow {}  // namespace arrow
+namespace arrow {
 
-#endif  // ARROW_TYPES_STRUCT_H
+class StructArray : public Array {
+ public:
+  StructArray(const TypePtr& type, int32_t length,
+      std::vector<ArrayPtr>& values,
+      int32_t null_count = 0,
+      std::shared_ptr<Buffer> null_bitmap = nullptr) :
+      Array(type, length, null_count, null_bitmap) {
+    type_ = type;
+    values_ = values;
+  }
+
+  virtual ~StructArray() {}
+
+  // Return a shared pointer in case the requestor desires to share ownership
+  // with this array.
+  const std::shared_ptr<Array>& values(int32_t pos) const {return values_.at(pos);}
+  const std::vector<ArrayPtr>& values() const {return values_;}
+
+  const std::shared_ptr<DataType>& value_type(int32_t pos) const {
+    return values_.at(pos)->type();
+  }
+
+  bool EqualsExact(const ListArray& other) const {
+    return true;
+  }
+  bool Equals(const std::shared_ptr<Array>& arr) const override {
+    return true;
+  }
+
+ protected:
+  // Contains kinds of Arrays.
+  std::vector<ArrayPtr> values_;
+};
+
+// ............................................................................
+// StrcutArray builder
+class StructBuilder : public ArrayBuilder {
+ public:
+  StructBuilder(MemoryPool* pool, const std::shared_ptr<DataType>& type,
+    const std::vector<FieldPtr>& fields,
+    std::vector<std::shared_ptr<ArrayBuilder>>& value_builder)
+    : ArrayBuilder(pool, type) {
+    fields_ = fields;
+    value_builder_ = value_builder;
+  }
+
+  Status Init(int32_t elements) {
+    return ArrayBuilder::Init(elements);
+  }
+
+  Status Resize(int32_t capacity) {
+    // Need space for the end offset
+    if (capacity < MIN_BUILDER_CAPACITY) {
+      capacity = MIN_BUILDER_CAPACITY;
+    }
+
+    if (capacity_ == 0) {
+      RETURN_NOT_OK(ArrayBuilder::Init(capacity));
+    } else {
+      RETURN_NOT_OK(ArrayBuilder::Resize(capacity));
+    }
+    capacity_ = capacity;
+
+    for (auto it = value_builder_.begin(); it != value_builder_.end(); ++it) {
+      it->get()->Resize(capacity);
+    }
+    return Status::OK();
+  }
+
+  // Vector append
+  //
+  // If passed, valid_bytes is of equal length to values, and any zero byte
+  // will be considered as a null for that slot
```

Comment:
Stale comment. This method must explain what action to take with the child builders to create a well-formed result